src/main/scala/io/github/kiemlicz/shelm/ChartDownloader.scala

```diff
@@ -33,18 +51,55 @@ object ChartDownloader {
           )
         else
           downloadDir / topDirs.head
-      case ChartLocation.AddedRepository(ChartName(name), ChartRepositoryName(repoName), chartVersion) =>
-        val options = s"$repoName/$name -d $downloadDir${chartVersion.map(v => s" --version $v").getOrElse("")} --untar"
-        IO.delete(downloadDir)
-        pullChart(options, sbtLogger)
-        downloadDir / name
-      case ChartLocation.RemoteRepository(ChartName(name), uri, settings, chartVersion) =>
-        val authOpts = HelmPlugin.chartRepositoryCommandFlags(settings)
-        val allOptions = s"--repo $uri $name $authOpts -d $downloadDir${chartVersion.map(v => s" --version $v").getOrElse("")} --untar"
-        IO.delete(downloadDir)
-        pullChart(allOptions, sbtLogger)
-        downloadDir / name
+      case ChartLocation.AddedRepository(chartName, ChartRepositoryName(repoName), chartVersion) =>
+        pullFromCacheOrRemote(
+          chartName, chartVersion, () => {
+            val options = s"$repoName/${chartName.name} -d $downloadDir${
+              chartVersion.map(v => s" --version $v").getOrElse("")
+            } --untar"
+            IO.delete(downloadDir)
+            pullChart(options, sbtLogger)
+          }
+        )
+      case ChartLocation.RemoteRepository(chartName, uri, settings, chartVersion) =>
+        pullFromCacheOrRemote(
+          chartName, chartVersion, () => {
+            val authOpts = HelmPlugin.chartRepositoryCommandFlags(settings)
+            val allOptions = s"--repo $uri ${chartName.name} $authOpts -d $downloadDir${
+              chartVersion.map(v => s" --version $v").getOrElse("")
+            } --untar"
+            IO.delete(downloadDir)
+            pullChart(allOptions, sbtLogger)
+          }
+        )
+    }
+  }
+
+  def calcCacheChartDir(
+    chartName: ChartName, chartVersion: Option[String], cacheBaseDir: File
+  ): File = chartVersion match {
+    case Some(version) => new File(cacheBaseDir, s"${chartName.name}-$version")
+    case None => new File(cacheBaseDir, s"${chartName.name}")
+  }
+
+  def pullFromCache(
+    chartName: ChartName,
+    chartVersion: Option[String],
+    downloadDir: File,
+    cacheBaseDir: File,
+  ): Option[File] = {
+    val chartDir = calcCacheChartDir(chartName, chartVersion, cacheBaseDir)
+    if (chartDir.isDirectory) {
+      IO.copyDirectory(chartDir, downloadDir / chartName.name, overwrite = true)
+      return Option(chartDir)
     }
+    Option.empty
+  }
+
+  def saveToCache(chartDir: File, chartName: ChartName, chartVersion: Option[String], cacheBaseDir: File): File = {
+    IO.copyDirectory(chartDir, cacheBaseDir, overwrite = true)
+    IO.move(cacheBaseDir / chartName.name, calcCacheChartDir(chartName, chartVersion, cacheBaseDir))
```

Comment:
Why do you do 2-stage copy-move? Can't we do a single copy without a move?

Reply:
refactored